sources/shell_launcher.c

```diff
@@ -140,6 +143,102 @@ sudo_closefrom(int lowfd) {
+   This is what we want, because all other BSD preparations here is done to create the
+   iTerm-Server to serve just that given user who started it. If another user is logged-in via
+   GUI via fast user switching, this other user will have it's own Per-User session, and will
+   have below it his Aqua per-session GUI belonging to it's UID, and his programs running there,
+   including his own iTerm-Server daemons forked just for it's UID use.
+   Security on MacOS is provided by BSD Kernel portion + Mach Kernel portion.
+   ----
+     System_Namespace [System]
+        |
+        ------ PerUSER_Namespace [Background] [user 501]
+        |      |
+        |      ----- PerSESSION_Namespace [Aqua] (GUI WindowServer) [user 501]
+        |
+        |
+        ------ PerUSER_Namespace [Background] [user 502]
+               |
+               ----- PerSESSION_Namespace [Aqua] (GUI WindowServer) [user 502]
+   ----
+
+   This way iTerm2 daemon is in conformation with current MacOS security rules mechanics,
+   and we completely eliminate all the problems that raised issue 4147. Per-user namespace is
+   the correct place for a user daemon to live and service it's services to the namespaces
+   levels below it.
+
+   *The Per-User namespace is the parent for all other context namespace for a given user. It
+   is never destroyed by GUI logout or crashes. And it will exist until there is at least one
+   daemon or service running on it, for that given user. [* from Apples documentation]
+
+   That's the exact namespace where Launchd places daemons for a given user.
+*/
+
+void MoveOutOfAquaSession() { // FIX for issue 4147
+    // This function must be called just before fork();
+    //
+    // This funcion move the process from the Aqua 'per-sesion' namespace to the
+    // same user 'per-user' namespace. This method is the standard method used
+    // by Apple for doing this kind of move. It is exactly the same move method
+    // used by Launchd.
+    mach_port_t parent = MACH_PORT_NULL;
+    kern_return_t kr;
+
+    kr = bootstrap_parent(bootstrap_port, &parent);
+
+    if (kr == KERN_SUCCESS) {
+        // Detach server daemon process from the user's Aqua 'per-session' namespace,
+        // and move it to the same UID 'per-user' namespace (just one level up).
+        mach_port_mod_refs(mach_task_self(), bootstrap_port, MACH_PORT_RIGHT_SEND, -1);
+        task_set_special_port(mach_task_self(), TASK_BOOTSTRAP_PORT, parent);
+
+        return; /* All Done! now return, as we are ready to begin forking! */
+    }
+    // Never reach. Unless if we had failed to get the parent bootstrap port, which according
+    // to Apple's documentation never happens. But I already saw rare wierd system situations
+    // where it may happen.
+    // Almost impossible, but if it happens, for sure is a BUG. So log it just in case.
+    FDLog(LOG_ERR,"BUG: Error getting parent BS port! Please report this msg and code: '%x'", kr);
```

Comment:
I'm a fan of defensive error logging

Reply:
Me too, learned the importance of that on my last 9 years working as developer on the asterisk telephony platform. There are a lot of similar BUG situations on that code, and naturally I created this function the way I'm accustomed to do. 

Thank you for your appreciation!